lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +104,149 @@ pub fn handle_completion(
         .id_of(params.text_document_position.text_document.uri.as_str())
         .unwrap();
 
+    let text = server.cache.files().source(file_id);
     let start = position_to_byte_index(
         server.cache.files(),
         file_id,
         &params.text_document_position.position,
     )
     .unwrap();
 
-    let locator = (file_id, ByteIndex(start as u32));
+    let locator = (file_id, ByteIndex((start - 1) as u32));
     let linearization = server.lin_cache_get(&file_id)?;
-
-    Trace::enrich(&id, linearization);
+    // lin_env is a mapping from idents to IDs.
+    // This is from the Linearizer used to build this linearization.
+    // idealy we should get the id and use it instead of comparing strings labels
+    let lin_env = &linearization.lin_env;
 
     let item = linearization.item_at(&locator);
-
     if item == None {
+        // NOTE: this might result in a case where completion is not working.
+        // This happens when an item cannot be found from the linearization.
         server.reply(Response::new_ok(id, Value::Null));
         return Ok(());
     }
-
     let item = item.unwrap().to_owned();
 
-    let in_scope: Vec<_> = linearization
-        .get_in_scope(&item)
+    Trace::enrich(&id, linearization);
+
+    let trigger = params.context.as_ref().and_then(|context| {
+        context
+            .trigger_character
+            .as_ref()
+            .map(|trigger| trigger.as_str())
+    });
+
+    let in_scope = match trigger {
+        // Record Completion
+        Some(".") => {
+            let name = get_identifier(&text[..start - 1])?;
```

Comment:
Another occurrence of `start - 1`.

I think this emphasizes my comment above: this probably deserves being extracted in a variable, and maybe a short comment on that variable to explain the `- 1`

Reply:
This one was to remove the `.`, but I don't think we need it anymore since `get_identifier` skips `.`s